### ISSUE TITLE
Add method to list project settings

### ISSUE THREAD

**Issue by winston-yallow** [godotengine/godot-proposals#8464]
### Describe the project you are working on

A plugin to configure input remapping

### Describe the problem or limitation you are having in your project

I am writing a plugin that allows developers to easily expose some input actions as remappable in-game to the end user. For this I need to show a list of possible input actions to expose.
There are also other proposals for example for getting a list of autoloads from the project settings: https://github.com/godotengine/godot-proposals/issues/4854
Also a discussion about this topic: https://github.com/godotengine/godot-proposals/discussions/4794

The most common workaround used in plugins at the moment is to use `get_property_list()` and iterate them. This does however not only return project settings but instead *all* properties.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Exposing a simple interface to get a list of settings allows to use them more conveniently. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

The idea is to expose a method that can return a list of project settings and optionally filter them:
```gdscript
func get_settings_list(filter := "") -> Array:
    if filter.is_empty:
        return all_settings
    return all_settings.filter(func(setting): setting.begins_with(filter))
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Yes, this can be worked around. However the workaround is not a very clean way to do this. It also depends on an implementation detail (settings being also properties). Exposing a dedicated method would decouple code accessing the settings list from the implementation detail.

### Is there a reason why this should be core and not an add-on in the asset library?

For the reasons stated above I think it should be a core feature. It's a super small change that shouldn't have any negative side effects, while making plugin development a bit nicer. I would be willing to implement this feature.

**Comment by winston-yallow** [comment]
I think it's clear, but just as an example how this can solve the other linked proposals too:
```gdscript
# Get list of autoloads
ProjectSettings.get_settings_list("autoload/")

# Get list of input actions
ProjectSettings.get_settings_list("input/")
```

**Comment by Calinou** [comment]
Should `filter` just be a `in` check or use `String.match()` (i.e. globbing)? Globbing would be more flexible.

**Comment by winston-yallow** [comment]
Good point, globbing makes this a lot more flexible. Could definitely be useful!

**Comment by KoBeWi** [comment]
> The most common workaround used in plugins at the moment is to use get_property_list() and iterate them. This does however not only return project settings but instead all properties.

If you want a list of only settings names, you can use
```GDScript
ProjectSettings.get_property_list().map(func(prop): return prop.name).filter(func(prop): return ProjectSettings.has_setting(prop))
```

**Comment by winston-yallow** [comment]
Yes I know, which is exactly why this workaround is (in my opinion) not a very clean way to do this. Especially since on the C++ side we already have the `props` as their own list, so it's trivial to expose them to GDScript.

**Comment by domdfcoding** [comment]
`get_property_list` doesn't appear to be documented in https://docs.godotengine.org/en/4.4/classes/class_projectsettings.html (or at least Ctrl+F doesn't find it). It's inherited from `Object` and is documented there, but it'd be handy to mention at the top where it talks about `get_setting`, `set_setting` and `has_setting`, as it's a logical counterpart to those but it currently isn't obvious it *can* be used to accomplish a similar goal.

**Comment by winston-yallow** [comment]
I don't think we should document it that way. The fact that project settings are stored as properties is an implementation detail and not something you should rely on.

Using it is just a workaround as the internal `props` property from C++ is not exposed to GDScript. That's why I propose to add a dedicated method to get the project settings.

**Comment by AThousandShips** [comment]
The engine itself uses properties to list these, and I'd say it's unlikely that this will change as it's the standard and proper system for all data like it

(To be clear it makes sense to add a feature to handle this specifically, but it's unlikely to ever be critical to not use the properties system IMO)
